File: lib/black_iron/pcs/pc.ex
defmodule BlackIron.PCs.PC do
  use Ecto.Schema
  import Ecto.Changeset

  @entype :pc

  @primary_key false

  @derive {Jason.Encoder,
           only: [
             :campaign_pid,
             :player_pid,
             :alias,
             :name,
             :description,
             :pronouns,
             :initiative,
             :portrait,
             :xp_added,
             :xp_spent
           ]}

  embedded_schema do
    field :__type__, :string
    field :campaign_pid, :string
    field :player_pid, :string
    field :alias, :string
    field :name, :string
    field :description, :string
    field :pronouns, :string
    field :initiative, Ecto.Enum, values: [:"out-of-combat", :"has-initiative", :"no-initiative"]
    field :portrait, :string
    field :xp_added, :integer
    field :xp_spent, :integer
  end

  def entype, do: @entype

  @doc false
  def changeset(pc, attrs \\ %{}) do
    pc
    |> cast(attrs, [
      :name,
      :description,
      :pronouns,
      :alias,
      :initiative,
      :portrait,
      :xp_added,
      :xp_spent
    ])
    |> validate_required([
      :name,
      :description,
      :pronouns,
      :alias,
      :initiative,
      :portrait,
      :xp_added,
      :xp_spent
    ])
    |> put_change(:__type__, to_string(@entype))
  end
end

File: assets/js/pcs/pc.ts
import convert from "url-slug";
import * as v from "valibot";

import { Campaign } from "../campaigns/campaign";
import { entity, IEntity } from "../entity";
import { mergeDeep } from "../utils/merge-deep";

export interface PCSchema {
  pcs: {
    key: string;
    value: IPC;
  };
}

export enum Initiative {
  OutOfCombat = "out-of-combat",
  HasInitiative = "has-initiative",
  NoInitiative = "no-initiative",
}

const pcDataSchema = v.object({
  campaign_pid: v.string(),
  alias: v.nullish(v.string()),
  name: v.string(),
  description: v.string(),
  pronouns: v.string(),
  initiative: v.enum(Initiative),
  portrait: v.nullish(v.string()),
  xp_added: v.pipe(
    v.unknown(),
    v.transform(Number),
    v.integer(),
    v.minValue(0),
  ),
  xp_spent: v.pipe(
    v.unknown(),
    v.transform(Number),
    v.integer(),
    v.minValue(0),
  ),
});

export type IPCData = v.InferOutput<typeof pcDataSchema>;

export type IPC = IEntity<IPCData>;

export class PC extends entity("pcs", pcDataSchema) {
  constructor(
    data: IPC,
    public campaign: Campaign,
    bumpRev?: boolean,
  ) {
    super(data);
    if (bumpRev) {
      this.bumpRev();
    }
  }

  get baseRoute() {
    return `/play/campaigns/${this.campaign ? this.campaign.pid : this.data.campaign_pid}/${
      convert(
        this.campaign.data.name,
      )
    }/pcs`;
  }

  get route() {
    return `${this.baseRoute}/${this.pid}/${convert(this.data.name)}`;
  }

  eq(other: IPC) {
    // TODO(@zkat): fill this in
    return super.eq(other);
  }

  merge(other: IPC) {
    return new PC(
      mergeDeep(this, other),
      this.campaign,
      true,
    ) as unknown as this;
  }
}
